File: examples/android-todo-app/app/src/main/kotlin/dev/pyric/example/todo/ui/screens/TodoScreen.kt
package dev.pyric.example.todo.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.Person
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import dev.pyric.example.todo.ui.components.AddTodoDialog
import dev.pyric.example.todo.ui.components.EmptyState
import dev.pyric.example.todo.ui.components.TodoFilterRow
import dev.pyric.example.todo.ui.components.TodoList
import dev.pyric.example.todo.ui.components.TodoTopAppBar
import dev.pyric.example.todo.ui.viewmodel.TodoViewModel

@Composable
fun TodoScreen(
    viewModel: TodoViewModel,
    modifier: Modifier = Modifier
) {
    val state by viewModel.uiState.collectAsStateWithLifecycle()
    val snackbarHostState = remember { SnackbarHostState() }
    var showAddDialog by remember { mutableStateOf(false) }

    LaunchedEffect(state.errorMessage) {
        state.errorMessage?.let { error ->
            snackbarHostState.showSnackbar(message = error)
            viewModel.clearError()
        }
    }

    Scaffold(
        modifier = modifier.fillMaxSize(),
        topBar = {
            TodoTopAppBar(
                isConnected = state.isConnected,
                onUnauthorizedWriteClick = viewModel::triggerUnauthorizedWrite
            )
        },
        snackbarHost = { SnackbarHost(hostState = snackbarHostState) },
        floatingActionButton = {
            if (state.isAuthenticated) {
                FloatingActionButton(
                    onClick = { showAddDialog = true }
                ) {
                    Icon(imageVector = Icons.Default.Add, contentDescription = "Add Todo")
                }
            }
        }
    ) { innerPadding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(innerPadding)
        ) {
            if (!state.isAuthenticated) {
                UnauthenticatedContent(
                    onSignInAnonymous = viewModel::signInAnonymously,
                    onSignInAlice = { viewModel.signInWithEmail() },
                    modifier = Modifier.weight(1f)
                )
            } else {
                UserHeaderBar(
                    userId = state.currentUserId.orEmpty(),
                    userEmail = state.currentUserEmail,
                    onSignOut = viewModel::signOut,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp, vertical = 6.dp)
                )

                TodoFilterRow(
                    selectedFilter = state.filter,
                    totalCount = state.totalCount,
                    activeCount = state.activeCount,
                    completedCount = state.completedCount,
                    onFilterSelected = viewModel::setFilter,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp, vertical = 4.dp)
                )

                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .weight(1f)
                ) {
                    when {
                        state.isLoading -> {
                            CircularProgressIndicator(
                                modifier = Modifier.align(Alignment.Center)
                            )
                        }
                        state.isFilterEmpty -> {
                            EmptyState(
                                filter = state.filter,
                                modifier = Modifier.align(Alignment.Center)
                            )
                        }
                        else -> {
                            TodoList(
                                todos = state.filteredTodos,
                                onToggle = viewModel::toggleTodo,
                                onDelete = viewModel::deleteTodo,
                                modifier = Modifier.fillMaxSize()
                            )
                        }
                    }
                }
            }
        }

        if (showAddDialog) {
            AddTodoDialog(
                onDismiss = { showAddDialog = false },
                onConfirm = { title ->
                    viewModel.addTodo(title)
                    showAddDialog = false
                }
            )
        }
    }
}

@Composable
private fun UserHeaderBar(
    userId: String,
    userEmail: String?,
    onSignOut: () -> Unit,
    modifier: Modifier = Modifier
) {
    Card(modifier = modifier) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 12.dp, vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(
                    imageVector = Icons.Default.Person,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.primary,
                    modifier = Modifier.size(20.dp)
                )
                Text(
                    text = "User: ${userEmail ?: userId.take(8)}",
                    style = MaterialTheme.typography.bodyMedium,
                    fontWeight = FontWeight.SemiBold,
                    modifier = Modifier.padding(start = 8.dp)
                )
            }
            OutlinedButton(
                onClick = onSignOut
            ) {
                Text("Sign Out", style = MaterialTheme.typography.labelSmall)
            }
        }
    }
}

@Composable
private fun UnauthenticatedContent(
    onSignInAnonymous: () -> Unit,
    onSignInAlice: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .fillMaxSize()
            .padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        Icon(
            imageVector = Icons.Default.Lock,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.primary,
            modifier = Modifier.size(64.dp)
        )
        Spacer(modifier = Modifier.height(16.dp))
        Text(
            text = "Authentication Required",
            style = MaterialTheme.typography.headlineSmall,
            fontWeight = FontWeight.Bold
        )
        Spacer(modifier = Modifier.height(8.dp))
        Text(
            text = "Firestore security rules enforce:\nallow read, write: if request.auth.uid == resource.data.userId;",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center
        )
        Spacer(modifier = Modifier.height(24.dp))
        Button(
            onClick = onSignInAnonymous,
            modifier = Modifier.fillMaxWidth(0.8f)
        ) {
            Text("Sign In Anonymously")
        }
        Spacer(modifier = Modifier.height(8.dp))
        OutlinedButton(
            onClick = onSignInAlice,
            modifier = Modifier.fillMaxWidth(0.8f)
        ) {
            Text("Sign In as Alice (Email/Password)")
        }
        Spacer(modifier = Modifier.height(16.dp))
        Text(
            text = "Or tap the Pyric Chip overlay floating above to 1-tap impersonate sandbox users or toggle Admin Bypass.",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.outline,
            textAlign = TextAlign.Center
        )
    }
}

File: examples/flutter-todo-app/lib/screens/todo_screen.dart
import 'package:flutter/material.dart';
import 'package:pyric_firestore/pyric_auth.dart';
import '../models/todo_item.dart';
import '../services/todo_repository.dart';

/// Primary Todo application screen featuring live stream, auth state, and rules diagnostics.
class TodoScreen extends StatefulWidget {
  final TodoRepository repository;

  const TodoScreen({
    super.key,
    required this.repository,
  });

  @override
  State<TodoScreen> createState() => _TodoScreenState();
}

class _TodoScreenState extends State<TodoScreen> {
  final TextEditingController _textController = TextEditingController();
  String? _effectiveUid;
  String? _userDisplay;

  @override
  void initState() {
    super.initState();
    _initAuthListener();
  }

  void _initAuthListener() {
    final auth = FirebaseAuthPlatform.instance;
    if (auth is PyricFirebaseAuthPlatform) {
      auth.authLensChanges.listen((lens) {
        if (!mounted) return;
        setState(() {
          _updateEffectiveIdentity(lens, auth.currentUser);
        });
      });
    }

    auth.authStateChanges().listen((user) {
      if (!mounted) return;
      setState(() {
        final lens = (auth is PyricFirebaseAuthPlatform)
            ? auth.currentAuthLens
            : AuthLens.anon;
        _updateEffectiveIdentity(lens, user);
      });
    });
  }

  void _updateEffectiveIdentity(AuthLens lens, UserPlatform? user) {
    if (lens.mode == 'as' && lens.uid != null) {
      _effectiveUid = lens.uid;
      _userDisplay = lens.uid;
    } else if (lens.mode == 'admin') {
      _effectiveUid = user?.uid ?? 'admin';
      _userDisplay = 'ADMIN';
    } else if (user != null) {
      _effectiveUid = user.uid;
      _userDisplay = user.email ?? user.displayName ?? user.uid;
    } else {
      _effectiveUid = null;
      _userDisplay = null;
    }
  }

  Future<void> _addTodo() async {
    final title = _textController.text.trim();
    if (title.isEmpty || _effectiveUid == null) return;
    _textController.clear();
    try {
      await widget.repository.addTodo(title, _effectiveUid!);
    } catch (e) {
      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text('Failed to add todo: $e')),
        );
      }
    }
  }

  Future<void> _triggerUnauthorizedWrite() async {
    try {
      await widget.repository.triggerUnauthorizedWrite();
    } catch (e) {
      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text('Rules check triggered (Expected): $e')),
        );
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    final auth = FirebaseAuthPlatform.instance;

    return Scaffold(
      appBar: AppBar(
        title: const Text('Pyric Flutter Todos'),
        actions: [
          IconButton(
            icon: const Icon(Icons.shield, color: Colors.redAccent),
            tooltip: 'Trigger Unauthorized Write (Verify Security Rules)',
            onPressed: _triggerUnauthorizedWrite,
          ),
        ],
      ),
      body: _effectiveUid == null
          ? _buildUnauthenticatedView(auth)
          : _buildAuthenticatedView(),
    );
  }

  Widget _buildUnauthenticatedView(FirebaseAuthPlatform auth) {
    return Center(
      child: Padding(
        padding: const EdgeInsets.all(24.0),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            const Icon(Icons.lock_outline, size: 64, color: Colors.orange),
            const SizedBox(height: 16),
            const Text(
              'Authentication Required',
              style: TextStyle(fontSize: 22, fontWeight: FontWeight.bold),
            ),
            const SizedBox(height: 8),
            const Text(
              'Firestore security rules require:\nallow read, write: if request.auth.uid == resource.data.userId;',
              textAlign: TextAlign.center,
              style: TextStyle(color: Colors.black54),
            ),
            const SizedBox(height: 24),
            ElevatedButton.icon(
              icon: const Icon(Icons.person),
              label: const Text('Sign In Anonymously'),
              onPressed: () async {
                try {
                  await auth.signInAnonymously();
                } catch (e) {
                  if (mounted) {
                    ScaffoldMessenger.of(context).showSnackBar(
                      SnackBar(content: Text('Sign-in failed: $e')),
                    );
                  }
                }
              },
            ),
            const SizedBox(height: 12),
            OutlinedButton.icon(
              icon: const Icon(Icons.email),
              label: const Text('Sign In as Alice (Email/Password)'),
              onPressed: () async {
                try {
                  await auth.signInWithEmailAndPassword('alice@example.com', 'password123');
                } catch (e) {
                  final msg = e.toString();
                  if (msg.contains('user-not-found') || msg.contains('No user found')) {
                    try {
                      await auth.createUserWithEmailAndPassword('alice@example.com', 'password123');
                      return;
                    } catch (ce) {
                      if (mounted) {
                        ScaffoldMessenger.of(context).showSnackBar(
                          SnackBar(content: Text('Sign-in failed: $ce')),
                        );
                      }
                      return;
                    }
                  }
                  if (mounted) {
                    ScaffoldMessenger.of(context).showSnackBar(
                      SnackBar(content: Text('Sign-in failed: $e')),
                    );
                  }
                }
              },
            ),
            const SizedBox(height: 16),
            const Text(
              'Or tap the Pyric Chip floating above to 1-tap impersonate sandbox users or toggle Admin Bypass.',
              textAlign: TextAlign.center,
              style: TextStyle(fontSize: 12, color: Colors.grey),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildAuthenticatedView() {
    return Column(
      children: [
        Container(
          padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
          color: Theme.of(context).colorScheme.surfaceContainerHighest,
          child: Row(
            children: [
              const Icon(Icons.account_circle, color: Colors.blue),
              const SizedBox(width: 8),
              Expanded(
                child: Text(
                  'User: $_userDisplay',
                  style: const TextStyle(fontWeight: FontWeight.bold),
                  overflow: TextOverflow.ellipsis,
                ),
              ),
              TextButton(
                onPressed: () => FirebaseAuthPlatform.instance.signOut(),
                child: const Text('Sign Out'),
              ),
            ],
          ),
        ),
        Padding(
          padding: const EdgeInsets.all(12.0),
          child: Row(
            children: [
              Expanded(
                child: TextField(
                  controller: _textController,
                  decoration: const InputDecoration(
                    hintText: 'What needs to be done?',
                    border: OutlineInputBorder(),
                    contentPadding: EdgeInsets.symmetric(horizontal: 12, vertical: 8),
                  ),
                  onSubmitted: (_) => _addTodo(),
                ),
              ),
              const SizedBox(width: 8),
              IconButton.filled(
                icon: const Icon(Icons.add),
                onPressed: _addTodo,
              ),
            ],
          ),
        ),
        Expanded(
          child: StreamBuilder<List<TodoItem>>(
            stream: widget.repository.getTodosStream(_effectiveUid!),
            builder: (context, snapshot) {
              if (snapshot.hasError) {
                return Center(
                  child: Text(
                    'Error: ${snapshot.error}',
                    style: const TextStyle(color: Colors.red),
                  ),
                );
              }
              if (snapshot.connectionState == ConnectionState.waiting) {
                return const Center(child: CircularProgressIndicator());
              }

              final todos = snapshot.data ?? [];
              if (todos.isEmpty) {
                return const Center(
                  child: Text(
                    'No todos yet for this user.\nAdd one above or switch identity via the chip!',
                    textAlign: TextAlign.center,
                    style: TextStyle(color: Colors.grey),
                  ),
                );
              }

              return ListView.builder(
                itemCount: todos.length,
                itemBuilder: (context, index) {
                  final item = todos[index];
                  return ListTile(
                    leading: Checkbox(
                      value: item.completed,
                      onChanged: (_) {
                        widget.repository.toggleTodo(item.id, item.completed);
                      },
                    ),
                    title: Text(
                      item.title,
                      style: TextStyle(
                        decoration: item.completed
                            ? TextDecoration.lineThrough
                            : TextDecoration.none,
                      ),
                    ),
                    trailing: IconButton(
                      icon: const Icon(Icons.delete_outline),
                      onPressed: () {
                        widget.repository.deleteTodo(item.id);
                      },
                    ),
                  );
                },
              );
            },
          ),
        ),
      ],
    );
  }
}
